### PR TITLE
add gym blackjack qlearning demo

### PR DESCRIPTION
**Pull request by brettkoonce**

### CODE REVIEW

File: Gym/Blackjack/main.swift [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        let priorQ = (1 - alpha) * oldQ
+
+        let maxReward = max(Q[post.playerSum][post.dealerCard][post.useableAce][0],
+                            Q[post.playerSum][post.dealerCard][post.useableAce][1])
+        let postQ = alpha * (Float(reward) + gamma * maxReward)
+
+        Q[prior.playerSum][prior.dealerCard][prior.useableAce][action] += priorQ + postQ
+    }
+
+    func getQLearningStrategy(observation: BlackjackState, iteration: Int) -> Bool {
```
Comment (rxwei) [inline Gym/Blackjack/main.swift]: The word "get" (or "set") is not quite idiomatic in Swift names. Swift functions should be named according to their side effects. Functions that are not side-effecting should read like nouns.  For example, this function should be named `qLearningStrategy(observation:iteration:)`. Same for all other occurrences of "get".

Source: [Swift API Design Guidelines - Strive for Fluent Usage](https://swift.org/documentation/api-design-guidelines/#strive-for-fluent-usage).

File: Gym/Blackjack/main.swift [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        let hitReward = Q[observation.playerSum][observation.dealerCard][observation.useableAce][0]
+        let stayReward = Q[observation.playerSum][observation.dealerCard][observation.useableAce][1]
+
+        if (iteration < Int.random(in: 1...learningPhase)) {
```
Comment (rxwei) [inline Gym/Blackjack/main.swift]:
```suggestion
        if iteration < Int.random(in: 1...learningPhase) {
```
Parentheses around conditions are redundant.

File: Gym/Blackjack/main.swift [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            return getRandomStrategy()
+        } else {
+            // quit learning after initial phase
+            if (iteration > learningPhase) { alpha = 0.0 }
```
Comment (rxwei) [inline Gym/Blackjack/main.swift]:
```suggestion
            if iteration > learningPhase { alpha = 0.0 }
```

File: Gym/Blackjack/main.swift [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+for solver in SolverType.allCases {
+    var totalReward = 0
+
+    for i in 1...totalIterations {
+        var isDone = false
+        environment.reset()
+
+        while !isDone {
+
```
Comment (rxwei) [inline Gym/Blackjack/main.swift]: Remove extra line.

File: Gym/Blackjack/main.swift (hunk 1 above)
Comment (brettkoonce) [reply Gym/Blackjack/main.swift]: kk, tweaked!

File: Gym/Blackjack/main.swift (hunk 4 above)
Comment (brettkoonce) [reply Gym/Blackjack/main.swift]: ✂️

File: Gym/Blackjack/main.swift (hunk 2 above)
Comment (brettkoonce) [reply Gym/Blackjack/main.swift]: removed!

File: Gym/Blackjack/main.swift (hunk 3 above)
Comment (brettkoonce) [reply Gym/Blackjack/main.swift]: removed!

File: Gym/Blackjack/main.swift [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    func normalStrategy(observation: BlackjackState) -> Bool {
+        if observation.playerSum == 0 {
+            return true
+        }
+        let lookupString = normalStrategyLookup(playerSum: observation.playerSum)
+        return Array(lookupString)[observation.dealerCard - 1] == "H"
+    }
+
+    func strategy(observation: BlackjackState, solver: SolverType, iteration: Int) -> Bool {
```
Comment (rxwei) [inline Gym/Blackjack/main.swift]: Hmm, what does the return type represent? The method name `strategy` indicates that it returns a strategy, but it seems to return a Boolean value with weak type information.

I'd suggest declaring a alias `typealias Strategy = Bool` and use `Strategy` where it makes sense.

Comment (brettkoonce) [reply Gym/Blackjack/main.swift]: tweaked, let me know what you think!

File: Gym/Blackjack/main.swift [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    let gamma: Float = 0.2
+
+    let numberOfPlayerStates = 32 // 21 + 10 + 1 offset
+    let numberOfDealerVisibleStates = 11 // 10 + 1 offset
+    let numberOfAceStates = 2 // useable / not bool
+    let numberOfPlayerActions = 2 // hit / stay
+
+    init() {
+        Q = Array(repeating: Array(repeating: Array(repeating: Array(repeating: 0.0,
+                                                                     count: numberOfPlayerActions),
```
Comment (rxwei) [inline Gym/Blackjack/main.swift]: One last thing: could you make sure everything fits within 100 columns?

Comment (brettkoonce) [reply Gym/Blackjack/main.swift]: 👻

File: Gym/Blackjack/main.swift [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+enum SolverType: CaseIterable {
+    case random, markov, qlearning, normal
+}
+
+class Solver {
+    var Q: [[[[Float]]]] = []
+    var alpha: Float = 0.5
+    let gamma: Float = 0.2
+
+    let numPlayerStates = 32 // 21 + 10 + 1 offset
```
Comment (rxwei) [inline Gym/Blackjack/main.swift]: Swift names do not use a "num" prefix to denote count. Instead, a singular noun followed by "count" is more idiomatic. As such, these names should instead be `playerStateCount`, `dealerVisibleStateCount`, `aceStateCount`, and `playerActionCount`.

Comment (brettkoonce) [reply Gym/Blackjack/main.swift]: 👍

File: Gym/Blackjack/main.swift [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Python
+import TensorFlow
+
+let gym = Python.import("gym")
+let environment = gym.make("Blackjack-v0")
+
+let totalIterations = 10000
```
Comment (rxwei) [inline Gym/Blackjack/main.swift]:
```suggestion
let totalIterations = 10000
```
```suggestion
let iterationCount = 10000
```

Comment (brettkoonce) [reply Gym/Blackjack/main.swift]: 🐯

**Comment by brettkoonce** [comment]
Docker is failing, step 7?  I can rebase/squash on latest master, if that helps.

**Comment by brettkoonce** [comment]
moved the tip, should work now!

**Comment by rxwei** [comment]
Thank you!

**Pull request merged**